### ISSUE TITLE
JWST datamodels does not handle FITS BINTABLE unsigned int32 correctly

### ISSUE THREAD

**Issue by philhodge** [spacetelescope/jwst#3179]
For the case that a column in a FITS BINTABLE is specified this way in the datamodels/schemas/:

      - name: DQ
        datatype: uint32

as in spec.schema.yaml, it appears that datamodels adds the TZEROi value twice.  (For uint32, TFORMi = 'J', and TZEROi = 2147483648.)  Here is an example, using a table written by the extract_1d step.  The header of the first extension of nrs1_bkg2_extract_1d.fits contains the following keywords:

    XTENSION= 'BINTABLE'           / binary table extension
    BITPIX  =                    8 / array data type
    NAXIS   =                    2 / number of array dimensions
    NAXIS1  =                   68 / length of dimension 1
    NAXIS2  =                 1382 / length of dimension 2
    PCOUNT  =                    0 / number of group parameters
    GCOUNT  =                    1 / number of groups
    TFIELDS =                    9 / number of table fields
    EXTNAME = 'EXTRACT1D'          / extension name
    EXTVER  =                    1 / extension value
    TTYPE1  = 'WAVELENGTH'
    TFORM1  = 'D       '
    TUNIT1  = 'um      '
    TTYPE2  = 'FLUX    '
    TFORM2  = 'D       '
    TUNIT2  = 'mJy     '
    TTYPE3  = 'ERROR   '
    TFORM3  = 'D       '
    TUNIT3  = 'mJy     '
    TTYPE4  = 'DQ      '
    TFORM4  = 'J       '
    TZERO4  =           2147483648
    TTYPE5  = 'NET     '
    TFORM5  = 'D       '
    TUNIT5  = 'DN/s    '
    TTYPE6  = 'NERROR  '
    TFORM6  = 'D       '
    TUNIT6  = 'DN/s    '
    TTYPE7  = 'BACKGROUND'
    TFORM7  = 'D       '
    TUNIT7  = 'DN/s    '
    TTYPE8  = 'BERROR  '
    TFORM8  = 'D       '
    TUNIT8  = 'DN/s    '
    TTYPE9  = 'NPIXELS '
    TFORM9  = 'D       '
    SLTNAME = 'S200A1  '           / Name of the slit
    SLITID  =                    0 / Slitlet ID
    SOURCEID=                    0 / Source ID
    SRCXPOS =                  0.0 / Source position in slit (x-axis)
    SRCYPOS =                  0.0 / Source position in slit (y-axis)
    SHUTSTA = '' / All (open and close) shutters in a slit
    SLIT_RA =    156.1911934350198 / Right ascension (deg) at middle of slit
    SLIT_DEC=   -45.68248321814713 / Declination (deg) at middle of slit
    SPORDER =                   -1 / Spectral order number
    END

Column 4 is the DQ column.

When I open this file using astropy.io.fits, the DQ values are zero:

    fd = fits.open("nrs1_bkg2_extract_1d.fits")
    dq = fd[1].data.field("dq")
    dq.min(), dq.max()
    (0, 0)

Note that this means the unscaled values in the DQ column are -2147483648, which gives 0 after adding TZERO4.

But when I open the file using datamodels, the DQ values are 2147483648:

    dm = datamodels.open("nrs1_bkg2_extract_1d.fits")
    type(dm)
    <class 'jwst.datamodels.multispec.MultiSpecModel'>
    dq = dm.spec[0].spec_table["dq"]
    dq.min(), dq.max()
    (2147483648, 2147483648)

**Comment by philhodge** [comment]
I temporarily changed multispec.schema.yaml and spec.schema.yaml to set the data type to uint16 (TZERO4 = 32768), and the problem persists:

    fd = fits.open("nrs1_bkg2_extract_1d.fits")
    dq = fd[1].data.field("dq")
    dq.dtype
    dtype('uint16')
    dq.min(), dq.max()
    (0, 0)

    dm = datamodels.open("nrs1_bkg2_extract_1d.fits")
    dq = dm.spec[0].spec_table["dq"]
    dq.min(), dq.max()
    (32768, 32768)

**Comment by jdavies-st** [comment]
Just making some notes here while its fresh in my mind.


The FITS `BinTableHDU` is opened via
https://github.com/spacetelescope/jwst/blob/6c8bd6b38e9636e07a1c5889387346225a6320b1/jwst/datamodels/fits_support.py#L561-L574

which passes it to
https://github.com/spacetelescope/jwst/blob/6c8bd6b38e9636e07a1c5889387346225a6320b1/jwst/datamodels/properties.py#L23-L48

And it the code then jumps over to `asdf` to map the schema to a numpy `dtype`
https://github.com/spacetelescope/asdf/blob/76e99ab242786f99fe1a74ece1524b7afd60d08a/asdf/tags/core/ndarray.py#L48-L86

It's somewhere in these last 2 functions that the mangling is occurring I think.

FWIW, I also think our WAVETAB problems are somewhere in these 2 functions as well.

**Comment by jdavies-st** [comment]
And the `dtype` generated by our schema is:
```
In [10]: datamodels.SpecModel() .spec_table.dtype                                                                                                                                              
Out[10]: dtype([('WAVELENGTH', '<f8'), ('FLUX', '<f8'), ('ERROR', '<f8'), ('DQ', '<u4'), ('NET', '<f8'), ('NERROR', '<f8'), ('BACKGROUND', '<f8'), ('BERROR', '<f8'), ('NPIXELS', '<f8')])
```

**Comment by hbushouse** [comment]
Fixed by #3680
